graf2d/gpadv7/inc/ROOT/RPadBase.hxx

```diff
@@ -81,43 +83,45 @@ public:
    /// \returns vector of vector (ret[x][y]) of created pads.
    std::vector<std::vector<std::shared_ptr<RPad>>> Divide(int nHoriz, int nVert, const RPadExtent &padding = {});
 
-   /// Create drawable of specified class T
-   template<class T, class... ARGS>
-   auto Draw(ARGS... args)
+   /// Add object to be painted.
+   /// Correspondent drawable will be created via GetDrawable() function which should be defined and be accessed at calling time.
+   /// If required, extra arguments for GetDrawable() function can be provided.
+   template <class T, class... ARGS>
+   auto Draw(const std::shared_ptr<T> &what, ARGS... args)
    {
-      auto drawable = std::make_shared<T>(args...);
+      // Requires GetDrawable(what) to be known!
+      auto drawable = GetDrawable(what, args...);
 
       TestIfFrameRequired(drawable.get());
 
-      fPrimitives.emplace_back(drawable);
+      AddPrimitive(drawable);
 
       return drawable;
    }
 
-   /// Add existing drawable instance to canvas
-   auto Draw(std::shared_ptr<RDrawable> &&drawable)
+   /// Create drawable of specified class T
+   template<class T, class... ARGS>
+   std::shared_ptr<T> Draw(ARGS... args)
    {
+      auto drawable = std::make_shared<T>(args...);
```

Comment:
This generic `T` might or might not be a `RDrawable`. As is, this interface is far too "open" / "permissive".

Reply:
Can we limit `T` only for derived from `RDrawable`?

In principle, it will complain anyway during compilation while `AddPrimitive(drawable)` expects only `RDrawable`

Reply:
Yes but it's extremely user unfriendly to understand what that flood of template diagnostics is talking about...